src/streams.rs

```diff
@@ -0,0 +1,698 @@
+    pub length: usize,
+    /// The very first entry in the stream.
+    pub first_entry: StreamId,
+    /// The very last entry in the stream.
+    pub last_entry: StreamId,
+}
+
+/// Reply type used with [`xinfo_consumer`] command, an array of every
+/// consumer in a specific consumer group.
+///
+/// [`xinfo_consumer`]: ../trait.Commands.html#method.xinfo_consumer
+///
+#[derive(Default, Debug, Clone)]
+pub struct StreamInfoConsumersReply {
+    /// An array of every consumer in a specific consumer group.
+    pub consumers: Vec<StreamInfoConsumer>,
+}
+
+/// Reply type used with [`xinfo_groups`] command.
+///
+/// This output represents all the consumer groups associated with
+/// the stream.
+///
+/// [`xinfo_groups`]: ../trait.Commands.html#method.xinfo_groups
+///
+#[derive(Default, Debug, Clone)]
+pub struct StreamInfoGroupsReply {
+    /// All the consumer groups associated with the stream.
+    pub groups: Vec<StreamInfoGroup>,
+}
+
+/// A consumer parsed from [`xinfo_consumers`] command.
+///
+/// [`xinfo_consumers`]: ../trait.Commands.html#method.xinfo_consumers
+///
+#[derive(Default, Debug, Clone)]
+pub struct StreamInfoConsumer {
+    /// Name of the consumer group.
+    pub name: String,
+    /// Number of pending messages for this specific consumer.
+    pub pending: usize,
+    /// This consumer's idle time in milliseconds.
+    pub idle: usize,
+}
+
+/// A group parsed from [`xinfo_groups`] command.
+///
+/// [`xinfo_groups`]: ../trait.Commands.html#method.xinfo_groups
+///
+#[derive(Default, Debug, Clone)]
+pub struct StreamInfoGroup {
+    /// The group name.
+    pub name: String,
+    /// Number of consumers known in the group.
+    pub consumers: usize,
+    /// Number of pending messages (delivered but not yet acknowledged) in the group.
+    pub pending: usize,
+    /// Last ID delivered to this group.
+    pub last_delivered_id: String,
+}
+
+/// Represents a pending message parsed from [`xpending`] methods.
+///
+/// [`xpending`]: ../trait.Commands.html#method.xpending
+#[derive(Default, Debug, Clone)]
+pub struct StreamPendingId {
+    /// The ID of the message.
+    pub id: String,
+    /// The name of the consumer that fetched the message and has
+    /// still to acknowledge it. We call it the current owner
+    /// of the message.
+    pub consumer: String,
+    /// The number of milliseconds that elapsed since the
+    /// last time this message was delivered to this consumer.
+    pub last_delivered_ms: usize,
+    /// The number of times this message was delivered.
+    pub times_delivered: usize,
+}
+
+/// Represents a stream `key` and its `id`'s parsed from `xread` methods.
+#[derive(Default, Debug, Clone)]
+pub struct StreamKey {
+    /// The stream `key`.
+    pub key: String,
+    /// The parsed stream `id`'s.
+    pub ids: Vec<StreamId>,
+}
+
+impl StreamKey {
+    /// Return only the stream `id`'s without their data.
+    pub fn just_ids(&self) -> Vec<&String> {
```

Comment:
This method seems a bit redundant. Remove it?

Reply:
This isn't redundant.  It's used to help reduce the tediousness of operating the `XCLAIM` command.  See `test_streams.rs`:

```rust
// claim and only return JUSTID
    let claimed: Vec<String> = con
        .xclaim_options(
            "k1",
            "g1",
            "c5",
            4,
            &claim_justids,
            StreamClaimOptions::default().with_force().with_justid(),
        )
        .unwrap();
```

Reply:
That call is made less annoying by using the mentioned method:

```rust
    let claim_justids = &reply.keys[0].just_ids();
```

Reply:
I haven't personally worked with it yet, but the `JUSTID` option is an [explicit part of Redis Streams API](
https://redis.io/commands/xclaim):

> Moreover, as a side effect, XCLAIM will increment the count of attempted deliveries of the message unless the JUSTID option has been specified (which only delivers the message ID, not the message itself). In this way messages that cannot be processed for some reason, for instance because the consumers crash attempting to process them, will start to have a larger counter and can be detected inside the system.

Reply:
Direct link to the existing test, for your browsing pleasure: https://github.com/Terkwood/redis-rs/blob/3f6cd92cf3f3f3e91fc3e0758dced3430c0ce38d/tests/test_streams.rs#L450

Reply:
Just wanted to check in on this thread.  If it helps us move toward a merge, I'm willing to play the devil's advocate against some of my statements above:  the JUSTIDs option for XCLAIM is more of an internally-focused subcommand, so I don't think there's a ton of necessity to include it _right this moment_.  From the docs:

> The command has multiple options, however most are mainly for internal use in order to transfer the effects of XCLAIM or other commands to the AOF file and to propagate the same effects to the slaves, and are unlikely to be useful to normal users

https://redis.io/commands/xclaim

So I think it could make sense to take any of the following paths, from an API design standpoint:

- go with the existing change set, which includes a nice helper method for JUSTIDs
- omit the JUSTIDs option for now, and debate its merits in a separate, future changeset (leaves us the option to add it later, in case it's desired by someone in the community)
- omit ALL of the internally-focused subcommands for XCLAIM, and raise a separate change set to add them later, as needed

Moving forward with any of these options will help us progress nicely, without painting us into a corner.

Thanks again for the thorough review!

Reply:
Thanks for bearing with me here. I do think this can be removed. Redis itself offers logic to give just the IDs back. Removing our wrapper logic from this PR doesn't hurt much. If someone does feel strongly about this we can consider adding it back in later.

I'll push an update to remove the just_ids wrapper, which should bring us to provisional consensus among the current PR participants.